p2p/service.py

```diff
@@ -136,9 +142,34 @@ def run_task(self, awaitable: Awaitable[Any]) -> None:
     def run_child_service(self, child_service: 'BaseService') -> None:
         """
         Run a child service and keep a reference to it to be considered during the cleanup.
+
+        If the child service finishes while we're still running, it will be restarted.
         """
-        self._child_services.append(child_service)
-        self.run_task(child_service.run())
+        self._child_services.add(child_service)
+
+        async def f() -> None:
+            try:
+                await child_service.run()
+            finally:
+                self._finished_child_services.put_nowait(child_service)
+
+        self.run_task(f())
+
+    async def _restart_finished_children(self) -> None:
+        restart_times: Dict[BaseService, List[float]] = {}
+        while self.is_running:
+            child = await self.wait(self._finished_child_services.get())
+            child_restart_times = restart_times.setdefault(child, [])
+            one_min_ago = time.time() - 60
+            if child_restart_times and max(child_restart_times) > one_min_ago:
+                self.logger.warning(
```

Comment:
It might make sense to `self.cancel()` here, as in most cases I believe a dead child will prevent the parent service from operating at all.